server/auth/types/jwt/jwt_auth.ts

```diff
@@ -0,0 +1,145 @@
+
+    const jwtToken = this.getJwtToken(request);
+    if (jwtToken) {
+      const authHeaderValue = `Bearer ${jwtToken}`;
+      let user: User;
+      try {
+        user = await this.securityClient.authenticateWithHeader(
+          request,
+          this.authHeaderName,
+          authHeaderValue
+        );
+      } catch (error) {
+        console.log(error);
+        return response.unauthorized();
+      }
+
+      const header: any = {};
+      header[this.authHeaderName] = authHeaderValue;
+
+      const cookie: SecuritySessionCookie = {
+        username: user.username,
+        credentials: {
+          authHeaderValue: authHeaderValue,
+        },
+        authType: JwtAuthentication.AUTH_TYPE,
+        expiryTime: Date.now() + this.config.cookie.ttl,
+      };
+      this.sessionStorageFactory.asScoped(request).set(cookie);
+
+      return tookit.authenticated({
+        requestHeaders: header,
+      });
+    }
+
+    // check if credentials present in cookie
+    let cookie = undefined;
+    try {
+      cookie = await this.sessionStorageFactory.asScoped(request).get();
+    } catch (error) {
+      console.log(error);
+    }
+
+    if (cookie) {
+      cookie.expiryTime = Date.now() + this.config.cookie.ttl;
+      this.sessionStorageFactory.asScoped(request).set(cookie);
+
+      const authHeaderValue = cookie.credentials.authHeaderValue;
+      if (authHeaderValue) {
+        const header: any = {};
+        header[this.authHeaderName] = authHeaderValue;
+        return tookit.authenticated({
+          requestHeaders: header,
+        });
+      }
+    }
+
+    // not credetail in query parameter, header, or cookie,
+    // redirect to login url if there is one
+    const loginEndpoint = this.config.jwt?.login_endpoint;
+    if (loginEndpoint) {
+      const loginUrl = parse(loginEndpoint, true);
+      let nextUrl: string = get(loginUrl.query, 'nexturl');
+      if (!nextUrl) {
+        nextUrl = format(request.url);
+        nextUrl = encodeURIComponent(nextUrl);
+        Object.assign(loginUrl.query, { nextUrl });
+      }
+      tookit.redirected({
+        location: format(loginUrl),
+      });
+    }
+
+    return tookit.notHandled(); // TODO: need to work with browser side to redirect to error page
+  };
+
+  private getJwtToken(request: KibanaRequest): string | undefined {
+    let token: string | undefined = undefined;
+
+    // try to extract JWT token from url query parameters
+    const urlParamName = this.config.jwt?.url_param;
+    if (urlParamName) {
+      token = (request.url.query as ParsedUrlQuery)[urlParamName] as string;
+      if (token) {
+        return token;
+      }
+    }
+
+    // fallback to HTTP header
+    const authHeaderValue = request.headers[this.authHeaderName];
+    if (authHeaderValue && !Array.isArray(authHeaderValue)) {
+      // authHeaderValue should not be array, to ensure client doesn't pass multiple auth header
```

Comment:
Could we move the comment above the corresponding code?

Reply:
moved